src/lib.rs

```diff
@@ -1,14 +1,363 @@
-pub fn add(left: u64, right: u64) -> u64 {
-    left + right
+use std::hash::{DefaultHasher, Hash, Hasher};
+
+/// Returns the hash of a single `Hash` value.
+fn hash_single<H: Hash>(value: H) -> u64 {
+    let mut hasher = DefaultHasher::new();
+    value.hash(&mut hasher);
+    hasher.finish()
+}
+
+/// Returns the hash resulting of combining two `Hash` values.
+fn hash_pair<H: Hash>(first: H, second: H) -> u64 {
+    let mut hasher = DefaultHasher::new();
+    first.hash(&mut hasher);
+    second.hash(&mut hasher);
+    hasher.finish()
+}
+
+/// Returns a given index's nth ancestor's index under its respective level.
+/// * `index` - The child index.
+/// * `level` - The relative upwards level of the target ancestor.
+fn ancestor_index(index: usize, level: usize) -> usize {
+    index / (2_usize.pow(level as u32)) // Integer division.
+}
+
+/// Returns the index of the sibling node of the given index.
+/// * `index` - The target index.
+fn sibling_index(index: usize) -> usize {
+    if index % 2 == 0 { index + 1 } else { index - 1 }
+}
+
+/// Given the starting leaves, generates all the upper levels of the tree
```

Comment:
What is a starting leaf? What is an upper level? Are upper levels related to ancestors?

Reply:
Updated.